File: top100/270-closest-binary-search-tree-value.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
 /*
 bst: 
    left: small, middle, right: larger
 */
public class Solution {
    public int closestValue(TreeNode root, double target) {
        int closest = root.val;
        double diff = Double.MAX_VALUE;
        
        //binary search 
        while (root != null){
            if (Math.abs(root.val - target)<diff){
                closest = root.val;
                diff = Math.abs(root.val - target);
            }
            
            if (root.val == target){
                return root.val;
            }
            else if (root.val<target){
                root = root.right;
            }
            else{
                root = root.left;
            }
        }
        
        return closest;
        
    }
}

File: leetcode/270.Closest Binary Search Tree Value.py
'''
Given a non-empty binary search tree and a target value, find the value in the BST that is closest to the target.

Note:
Given target value is a floating point.
You are guaranteed to have only one unique value in the BST that is closest to the target.
'''
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def closestValue(self, root, target):
        """
        :type root: TreeNode
        :type target: float
        :rtype: int
        """
        ret = root.val 
#         traverse the tree
        while root:
            if abs(root.val-target) < abs(ret-target):
                ret = root.val
            
            root = root.left if target < root.val else root.right
        
        return ret
        
